### PR TITLE
feat: introduce when expressions to steps

### PR DESCRIPTION
**Pull request by ericzzzzzzz**
# Changes
This PR introduces the use of "when" expressions within Tekton steps, enabling conditional execution. The controller marshals the taskSpec.steps[*].when data and passes it to the entrypointer. The entrypointer evaluates these expressions to determine if a step should run. Skipped steps exit with code 0 and a 'Skipped' TerminationReason. This approach allows steps to use previous step results in their conditions, even though the controller still creates the container for skipped steps.

Note: StepAction doesn't own whenExpressions, whenExpressions should be defined under `steps[*]` to control the flow. 

# fixes: #7623

## Related 
 - tep: https://github.com/tektoncd/community/pull/1147 

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] [pre-commit](https://github.com/tektoncd/pipeline/blob/main/DEVELOPMENT.md#install-tools) Passed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
 introduce when expressions to steps.
```

/kind feat

### CODE REVIEW

File: config/config-feature-flags.yaml [hunk 1]
```diff
@@ -125,7 +125,7 @@ data:
   enable-cel-in-whenexpression: "false"
   # Setting this flag to "true" will enable the use of StepActions in Steps
   # This feature is in preview mode and not implemented yet. Please check #7259 for updates.
-  enable-step-actions: "false"
+  enable-step-actions: "true"
```
Comment (chitrangpatel) [inline config/config-feature-flags.yaml:128]: disable this when submitting the final PR :) before removing the hold/draft.

Comment (ericzzzzzzz) [reply config/config-feature-flags.yaml:128]: sure, thank you for the reminding

File: go.mod [hunk 2]
```diff
@@ -1,6 +1,8 @@
 module github.com/tektoncd/pipeline
 
-go 1.19
+go 1.21
+
+toolchain go1.21.0
```
Comment (chitrangpatel) [inline go.mod:5]: Did this come from running `./hack/update-codegen.sh`? If not, may be we can rever this and let dependabots bump it for us.

**Review by chitrangpatel — COMMENTED** [review]
I think we need conversion for `When` between `v1` and `v1beta1`.

**Review by chitrangpatel — APPROVED** [review]

**Comment by chitrangpatel** [comment]
Thanks for your work here @ericzzzzzzz
/meow

**Comment by chitrangpatel** [comment]
@tektoncd/core-collaborators PTAL 🙏 
It unlocks https://github.com/tektoncd/pipeline/issues/7623.

File: pkg/entrypoint/entrypointer.go [hunk 3]
```diff
@@ -68,6 +70,8 @@ var (
 	ErrContextCanceled = ContextError(context.Canceled.Error())
 	// ErrSkipPreviousStepFailed is the error returned when the step is skipped due to previous step error
 	ErrSkipPreviousStepFailed = SkipError("error file present, bail and skip the step")
+
+	ErrSkipWhenExpressionsFalse = SkipError("error file present, bail and skip the step")
```
Comment (khrm) [inline pkg/entrypoint/entrypointer.go:74]: nit: This isn't needed.

**Comment by chitrangpatel** [comment]
@tektoncd/core-collaborators looking for a non-googler review on this 🙏

**Review by JeromeJu — APPROVED** [review]

**Review by vdemeester — COMMENTED** [review]
Naive question : we didn't have when expression for steps, did we ?

**Comment by chitrangpatel** [comment]
> Naive question : we didn't have when expression for steps, did we ?

No, we didnt. That's likely because we didn't have things like StepResults and also everything was inlined so Task authors could add control logic in the Steps. 

With StepActions, it's difficult to do that since users cannot really change the logic in a StepAction so when expressions do help.

**Comment by vdemeester** [comment]
Essentially, we are adding a "pipeline" behavior (skip a task) to tasks (with steps). And we are not going through a TEP 😓  I know this is *touched* [here](https://github.com/tektoncd/community/blob/main/teps/0142-enable-step-reusability.md#conditional-execution-of-steps) but I would love to have more detail and exploration around this before committing to it.

A question, in the example we have, we use the "old" syntax, don't we ? See [TEP-0145](https://github.com/tektoncd/community/blob/main/teps/0145-cel-in-whenexpression.md) that seems implemented.

A nit, this doesn't introduce when expressions to stepActions but to steps (and thus, stepactions) 👼🏼

**Comment by chitrangpatel** [comment]
> Essentially, we are adding a "pipeline" behavior (skip a task) to tasks (with steps). And we are not going through a TEP 😓 I know this is _touched_ [here](https://github.com/tektoncd/community/blob/main/teps/0142-enable-step-reusability.md#conditional-execution-of-steps) but I would love to have more detail and exploration around this before committing to it.
> 
Thats fair. The reason we didn't think that a TEP was necessary was because we were basically implementing the same thing at the Step level (I think similar to https://github.com/tektoncd/pipeline/pull/7100). Apart from the fact that there are when expressions around Steps, there is nothing novel in technique that requires a new design. We can definitely wrap it in a short TEP or an issue, whatever we think is a good way forward.

> A question, in the example we have, we use the "old" syntax, don't we ? See [TEP-0145](https://github.com/tektoncd/community/blob/main/teps/0145-cel-in-whenexpression.md) that seems implemented.
> 

We also allow CEL here. Please keep me honest here @ericzzzzzzz.

> A nit, this doesn't introduce when expressions to stepActions but to steps (and thus, stepactions) 👼🏼
Agreed!

**Comment by ericzzzzzzz** [comment]
> We also allow CEL here. Please keep me honest here @ericzzzzzzz.

Yes, we support both regular when expressions and `cel`.  My understanding is that they're different approaches to evaluation conditions, `cel` is not something meant to replace original `when Expressions`  right?

**Comment by ericzzzzzzz** [comment]
Hi @vdemeester [tep-0156](https://github.com/tektoncd/community/pull/1147) is merged now, could you another look at this pr? Thanks

File: docs/stepactions.md [hunk 4]
```diff
@@ -521,3 +522,104 @@ spec:
 ```
 
 The default resolver type can be configured by the `default-resolver-type` field in the `config-defaults` ConfigMap (`alpha` feature). See [additional-configs.md](./additional-configs.md) for details.
+
+### Controlling Step Execution with when Expressions
+
+You can define `when` in a StepAction to control its execution.
```
Comment (vdemeester) [inline docs/stepactions.md:528]: Isn't this wrong ? You can define `when` in a step (that refers a `StepAction`), but you cannot define `when` in the `StepAction` itself, can you ?

Comment (ericzzzzzzz) [reply docs/stepactions.md:528]: ahh! You're right!  This should be a `step`

Comment (ericzzzzzzz) [reply docs/stepactions.md]: updated

**Comment by ericzzzzzzz** [comment]
Hi @vdemeester , I updated the doc, please take another look, if you have any other questions, let me know! Thanks!

**Pull request merged**